File: backend/messenger.js
var Message = /** @class */ (function () {
    function Message(sender, content, timestamp) {
        this.sender = sender;
        this.content = content;
        this.timestamp = timestamp;
    }
    // __repr__ equivalent
    Message.prototype.toString = function () {
        return "Message from ".concat(this.sender, " at ").concat(this.timestamp, ": ").concat(this.content);
    };
    return Message;
}());
var Conversation = /** @class */ (function () {
    function Conversation(participants, product) {
        this.participants = participants;
        this.product = product;
        this.messages = [];
    }
    Conversation.prototype.addMessage = function (sender, content) {
        if (this.participants.includes(sender)) {
            var now = new Date().toISOString().slice(0, 19).replace('T', ' ');
            var message = new Message(sender, content, now);
            this.messages.push(message);
            return true;
        }
        return false;
    };
    // __repr__ equivalent
    Conversation.prototype.toString = function () {
        var output = '';
        this.messages.forEach(function (message) {
            output += "".concat(message.toString(), " \n");
        });
        return "<Conversation between ".concat(this.participants.join(', '), ">\n").concat(output);
    };
    return Conversation;
}());
var Messenger = /** @class */ (function () {
    function Messenger() {
        this.conversations = [];
    }
    Messenger.prototype.addConversation = function (conversation) {
        this.conversations.push(conversation);
    };
    Messenger.prototype.getConversations = function () {
        return this.conversations;
    };
    // __repr__ equivalent
    Messenger.prototype.toString = function () {
        var output = '';
        this.conversations.forEach(function (conversation) {
            output += "".concat(conversation.toString(), " \n");
        });
        return "<Messenger with ".concat(this.conversations.length, " conversations>,\n").concat(output);
    };
    return Messenger;
}());
// Test Example
var messenger = new Messenger();
var convo1 = new Conversation(["John", "Kim"], "Book");
messenger.addConversation(convo1);
convo1.addMessage("John", "Book please?");
convo1.addMessage("Kim", "FS $20");
convo1.addMessage("John", "Great");
var convo2 = new Conversation(["Bill", "Kate"], "Calculator");
messenger.addConversation(convo2);
convo2.addMessage("Bill", "Calc please?");
convo2.addMessage("Kate", "By calc do you mean calculator");
convo2.addMessage("Bill", "Yup");
console.log(messenger.toString());

File: backend/messenger.py
from datetime import datetime

class Message:
    def __init__(self, sender: str, content: str, timestamp: str):
        self.sender = sender
        self.content = content
        self.timestamp = timestamp

    def __repr__(self):
        return f"Message from {self.sender} at {self.timestamp}: {self.content}"

class Conversation:
    def __init__(self, participants: list[str], product):
        self.participants = participants
        self.product = product
        self.messages: list[Message] = []

    def add_message(self, sender, content):
        if (self.participants.__contains__(sender)):
            now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            message = Message(sender, content, now)
            self.messages.append(message)
            return True
        else:
            return False

    def __repr__(self):
        output = ""
        for message in self.messages:
            output = f"{output} {message} \n"
        return f"<Conversation between {', '.join(self.participants)}> \n{output}"

class Messenger:
    def __init__(self):
        self.conversations: list[Conversation] = []

    def add_conversation(self, conversation: Conversation):
        self.conversations.append(conversation)

    def get_conversations(self):
        return self.conversations

    def __repr__(self):
        output = ""
        for conversation in self.conversations:
            output = f"{output} {conversation} \n"
        return f"<Messenger with {len(self.conversations)} conversations>, \n{output}"
    
messenger = Messenger()
convo = Conversation(["John", "Kim"], "Book")
messenger.add_conversation(convo)
convo.add_message("John", "Book please?")
convo.add_message("Kim", "FS $20")
convo.add_message("John", "Great")
convo = Conversation(["Bill", "Kate"], "Calculator")
messenger.add_conversation(convo)
convo.add_message("Bill", "Calc please?")
convo.add_message("Kate", "By calc do you mean calculator")
convo.add_message("Bill", "Yup")
print(messenger)